README.md
# go-mesos-framework-kafka

[![Donate](https://img.shields.io/liberapay/receives/AVENTER.svg?logo=liberapay)](https://liberapay.com/mesos)
[![Support Chat](https://img.shields.io/static/v1?label=Chat&message=Support&color=brightgreen)](https://riot.im/app/#/room/#support:matrix.aventer.biz)

Dies ist ein Kafka Framework für Apache Mesos

## Voraussetzung

- Apache Mesos ab 1.6.0
- Mesos mit SSL und Authentication ist Optional
- Persistent Storage

## Framework starten

Mit den folgenden Umgebungsvariablen kann das Framework konfiguriert werden. Nach dem Starten wird es sich an den Mesos Master anmelden und erscheint als "kafkaframework" in der Mesos UI. Sobald sich das Framework erfolgreich initialisiert hat, startet es Zookeeper und anschließend Kafka.



```Bash
export FRAMEWORK_USER="root"
export FRAMEWORK_NAME="kafkaframework"
export FRAMEWORK_PORT="10000"
export FRAMEWORK_ROLE="kafka"
export FRAMEWORK_STATEFILE_PATH="/tmp"
export MESOS_PRINCIPAL="<mesos_principal>"
export MESOS_USERNAME="<mesos_user>"
export MESOS_PASSWORD="<mesos_password>"
export MESOS_MASTER="<mesos_master_server>:5050"
export LOGLEVEL="DEBUG"
export DOMAIN="weave.local"
export ZOOKEEPER_COUNT=1
export KAFKA_COUNT=3
export RES_CPU=0.1
export RES_MEM=3200
export AUTH_PASSWORD="password"
export AUTH_USERNAME="user"
export MESOS_SSL="true"
export ZOOKEEPER_CUSTOM_DOMAIN=""
export KAFKA_CUSTOM_DOMAIN=""
export IMAGE_KAFKA="confluentinc/cp-kafka:5.4.1"
export IMAGE_ZOOKEEPER="zookeeper:3.5.7"
export VOLUME_DRIVER="local"
export VOLUME_KAFKA="/tmp/kafka1,/tmp/kafka2,/tmp/kafka3"
export VOLUME_ZOOKEEPER="/tmp/zookeeper1,/tmp/zookeeper2,/tmp/zookeeper3"

go run init.go app.go
```

![Kafka Framework in Mesos](kafka_mesos.gif)

Faellt ein Container aus, wird dieser neugestartet.

![Kafka Framework in Mesos](kafka_mesos1.gif)

## Task Status Abfragen

Um den Status eines Tasks über das Framework abzufragen, folgendes Kommando verwenden:

```Bash
curl -X GET 127.0.0.1:10000/v0/container/<taskId> -d 'JSON'  | jq
```

## Fehlende Kafka oder Zookeeper Starten

Sollte aus bestimmten Gründen der Healthcheck Status im Framework nicht mit der Realität übereinstimmen, kann über den nachfolgenden Aufruf erzwungen werden die fehlenden Container zu starten.

```Bash
curl -X GET 127.0.0.1:10000/v0/<kafka|zookeeper>/reflate -d 'JSON'
```

## Kafka oder Zookeeper skalieren

Um Kafka oder Zookeeper im Betrieb zu skalieren, wird dem Framework die zu laufende Anzahl an Containern angegeben. Soll der Zookeeper also drei mal laufen, muss als <count> eine "3" angegeben werden. Beim Scaledown wird der zuletzt hinzugefügte Container entfernt.

```Bash
curl -X GET 127.0.0.1:10000/v0/<kafka|zookeeper>/scale/<count> -d 'JSON'
```

## Task killen

Sollte es notwendig sein einen Task zu benden, erfolgt dies mit dem nachfolgenden aufruf. Der beendete Container wird nicht automatisch neu gestartet.

```Bash
curl -X GET 127.0.0.1:10000/v0/task/kill/<taskId> -d 'JSON'
```


changelog.md
# Changelog

## master

- Fix scale down mismatch
- Removes via api killed tasks from config. Same with scale down agents/servers. 

## v0.0.4

- Add volume driver support

## v0.0.3

- Add persistent storage Support
- Change to go mod

## v0.0.2

- Add start mesos container
- Add start command
- Add persist framework info therefore the framework know what to do after a crash
- Add save task state
- Add start kafka and zookeeper
- Add persist task state
- Add multinode support
- Add container monitor to restart if a container failed
- Add RestCall to reflate missing zookeeper and kafka processes, for the case the monitoring can not find the problem
- Add RestCall scale up and down for zookeeper
- Add RestCall to kill a task
- Add Authentication
- Add Support to configure (non)SSL Support to Mesos
- Add Custom Domain to Zookeeper and Kafka to match Consul DNS
- Add Service Name ENV Variable to Zookeeper and Kafka to match Consul DNS
- Add Call_Suppress to tell mesos it does not send us offers until we ask
- Add default values for some env
- Add custom image name via env


init.go
package main

import (
	"os"
	"strconv"
	"strings"

	util "git.aventer.biz/AVENTER/util.git"

	cfg "go-mesos-framework-kafka/types"
)

var config cfg.Config

func init() {
	config.ZookeeperMax = 0
	config.KafkaMax = 0
	config.ZookeeperCount = 0
	config.KafkaCount = 0

	config.FrameworkUser = util.Getenv("FRAMEWORK_USER", "root")
	config.FrameworkName = util.Getenv("FRAMEWORK_NAME", "kafka")
	config.FrameworkRole = util.Getenv("FRAMEWORK_ROLE", "kafka")
	config.FrameworkPort = util.Getenv("FRAMEWORK_PORT", "10000")
	config.FrameworkInfoFilePath = util.Getenv("FRAMEWORK_STATEFILE_PATH", "/tmp")
	config.Principal = os.Getenv("MESOS_PRINCIPAL")
	config.Username = os.Getenv("MESOS_USERNAME")
	config.Password = os.Getenv("MESOS_PASSWORD")
	config.MesosMasterServer = os.Getenv("MESOS_MASTER")
	config.LogLevel = util.Getenv("LOGLEVEL", "info")
	config.Domain = os.Getenv("DOMAIN")
	config.ZookeeperMax, _ = strconv.Atoi(os.Getenv("ZOOKEEPER_COUNT"))
	config.KafkaMax, _ = strconv.Atoi(os.Getenv("KAFKA_COUNT"))
	config.ResCPU, _ = strconv.ParseFloat(util.Getenv("RES_CPU", "0.1"), 64)
	config.ResMEM, _ = strconv.ParseFloat(util.Getenv("RES_MEM", "1200"), 64)
	config.Credentials.Username = os.Getenv("AUTH_USERNAME")
	config.Credentials.Password = os.Getenv("AUTH_PASSWORD")
	config.AppName = "Mesos Kafka Framework"
	config.ZookeeperCustomString = os.Getenv("ZOOKEEPER_CUSTOM_DOMAIN")
	config.KafkaCustomString = os.Getenv("KAFKA_CUSTOM_DOMAIN")
	config.ImageZookeeper = util.Getenv("IMAGE_ZOOKEEPER", "zookeeper:3.5.7")
	config.ImageKafka = util.Getenv("IMAGE_KAFKA", "confluentinc/cp-kafka:5.4.1")
	config.VolumeDriver = util.Getenv("VOLUME_DRIVER", "local")
	config.VolumeKafka = util.SplitString(util.Getenv("VOLUME_KAFKA", "/data/kafka,"))
	config.VolumeZookeeper = util.SplitString(util.Getenv("VOLUME_ZOOKEEPER", "/data/zookeeper,"))

	// The comunication to the mesos server should be via ssl or not
	if strings.Compare(os.Getenv("MESOS_SSL"), "true") == 0 {
		config.MesosSSL = true
	} else {
		config.MesosSSL = false
	}

}
